File: 3413.maximum-coins-from-k-consecutive-bags.py
#
# @lc app=leetcode id=3413 lang=python3
#
# [3413] Maximum Coins From K Consecutive Bags
#

from typing import List
from collections import defaultdict

# @lc code=start
class Solution:
    def maximumCoins(self, coins: List[List[int]], k: int) -> int:
        # Map to store total coins at each position
        position_coins = defaultdict(int)
        
        # Fill map with coin values from segments
        for li, ri, ci in coins:
            for pos in range(li, ri + 1):
                position_coins[pos] += ci
        
        # Sort positions for sliding window traversal
        sorted_positions = sorted(position_coins.keys())
        
        max_coins = 0
        current_sum = 0
        left = 0
        
        # Use sliding window over sorted positions to find max sum of k consecutive positions
        for right in range(len(sorted_positions)):
            current_sum += position_coins[sorted_positions[right]]
            
            if right - left + 1 > k:
                current_sum -= position_coins[sorted_positions[left]]
                left += 1
            
            if right - left + 1 == k:
                max_coins = max(max_coins, current_sum)
                
        return max_coins
# @lc code=end

File: 3413.maximum-coins-from-k-consecutive-bags.java
#
# @lc app=leetcode id=3413 lang=java
#
# [3413] Maximum Coins From K Consecutive Bags
#
# @lc code=start
class Solution {
    public long maximumCoins(int[][] coins, int k) {
        Arrays.sort(coins, (a, b) -> Integer.compare(a[0], b[0]));
        int n = coins.length;
        int left = 0, right = 0;
        long sum = 0, max = 0;
        long windowStart = coins[0][0];
        long windowEnd = windowStart + k - 1;

        // precompute the total number of bags
        while (right < n && coins[right][1] < windowStart) right++;

        long l = windowStart;
        while (true) {
            // Reset sum for each window
            sum = 0;
            long curL = l, curR = l + k - 1;
            int idx = 0;
            while (idx < n && coins[idx][1] < curL) idx++;
            int j = idx;
            while (j < n && coins[j][0] <= curR) {
                long segL = Math.max(coins[j][0], curL);
                long segR = Math.min(coins[j][1], curR);
                if (segL <= segR) {
                    sum += (segR - segL + 1) * coins[j][2];
                }
                j++;
            }
            max = Math.max(max, sum);
            // Move window forward
            // Find next possible l: the next start of any interval or current l + 1, whichever is minimal, but not beyond possible window range
            long nextL = l + 1;
            if (idx < n) {
                nextL = Math.max(nextL, Math.min(coins[idx][1] + 1, l + 1));
            }
            // Stop if window exceeds range
            if (j >= n && (idx >= n || coins[n - 1][1] < nextL)) break;
            l = nextL;
        }
        return max;
    }
}
# @lc code=end